File: runner.py
#!/usr/bin/env python3
"""
mtr-runner: periodically runs mtr against configured destinations
and saves JSON output to a configurable path.
"""

import json
import os
import subprocess
import time
from datetime import datetime, timezone
from pathlib import Path


def get_env(key: str, default: str) -> str:
    return os.environ.get(key, default).strip()


def sanitize(dest: str) -> str:
    return dest.replace(".", "-").replace(":", "-").replace("/", "-")


def run_mtr(mtr_bin: str, dest: str, count: int, output_path: Path) -> None:
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    outfile = output_path / f"{timestamp}_{sanitize(dest)}.json"

    print(f"[mtr-runner] Running mtr to {dest} -> {outfile}", flush=True)

    try:
        result = subprocess.run(
            [mtr_bin, "-j", "-c", str(count), dest],
            capture_output=True,
            text=True,
            timeout=count * 5 + 30,  # generous timeout
        )

        if result.returncode != 0:
            print(f"[mtr-runner] WARNING: mtr exited {result.returncode} for {dest}", flush=True)

        # Try to parse as JSON to validate, fall back to raw output
        raw = result.stdout.strip() if result.stdout.strip() else result.stderr.strip()
        try:
            data = json.loads(raw)
            outfile.write_text(json.dumps(data, indent=2))
        except json.JSONDecodeError:
            outfile.write_text(raw)
            print(f"[mtr-runner] WARNING: Output for {dest} was not valid JSON", flush=True)

    except subprocess.TimeoutExpired:
        print(f"[mtr-runner] ERROR: mtr timed out for {dest}", flush=True)
    except Exception as e:
        print(f"[mtr-runner] ERROR: {e}", flush=True)


def main() -> None:
    interval = int(get_env("MTR_INTERVAL", "300"))
    count = int(get_env("MTR_COUNT", "10"))
    output_path = Path(get_env("MTR_OUTPUT_PATH", "/data/mtr"))
    destinations_raw = get_env("MTR_DESTINATIONS", "1.1.1.1")
    mtr_bin = get_env("MTR_BIN", "/usr/sbin/mtr")

    destinations = [d.strip() for d in destinations_raw.split(",") if d.strip()]

    output_path.mkdir(parents=True, exist_ok=True)

    print(f"[mtr-runner] Starting.", flush=True)
    print(f"  Interval : {interval}s", flush=True)
    print(f"  Count    : {count} packets", flush=True)
    print(f"  Output   : {output_path}", flush=True)
    print(f"  Targets  : {destinations}", flush=True)

    while True:
        for dest in destinations:
            run_mtr(mtr_bin, dest, count, output_path)

        print(f"[mtr-runner] Cycle complete. Sleeping {interval}s...", flush=True)
        time.sleep(interval)


if __name__ == "__main__":
    main()

File: runner.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

func getEnv(key, def string) string {
	v := strings.TrimSpace(os.Getenv(key))
	if v == "" {
		return def
	}
	return v
}

func sanitize(dest string) string {
	r := strings.NewReplacer(".", "-", ":", "-", "/", "-")
	return r.Replace(dest)
}

func runMTR(mtrBin, dest string, count int, outputPath string) {
	ts := time.Now().UTC().Format("20060102T150405Z")
	outfile := filepath.Join(outputPath, fmt.Sprintf("%s_%s.json", ts, sanitize(dest)))
	fmt.Printf("[mtr-runner] Running mtr to %s -> %s\n", dest, outfile)

	// mtr 0.96 doesn't support JSON, use CSV instead
	cmd := exec.Command(mtrBin, "-r", "-c", fmt.Sprintf("%d", count), "-n", dest)
	out, err := cmd.Output()
	if err != nil {
		fmt.Printf("[mtr-runner] ERROR: mtr failed for %s: %v\n", dest, err)
		return
	}

	// Parse CSV output and convert to JSON
	data := parseMTRCSV(string(out))
	pretty, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(outfile, pretty, 0644)
}

func parseMTRCSV(csv string) map[string]interface{} {
	lines := strings.Split(strings.TrimSpace(csv), "\n")
	if len(lines) < 1 {
		return map[string]interface{}{"error": "no output"}
	}

	// Parse header
	headers := strings.Fields(strings.TrimSpace(lines[0]))
	if len(headers) < 8 {
		return map[string]interface{}{"error": "invalid output format"}
	}

	// Parse data rows
	var hops []map[string]interface{}
	for i := 1; i < len(lines); i++ {
		fields := strings.Fields(strings.TrimSpace(lines[i]))
		if len(fields) < len(headers) {
			continue
		}

		hop := make(map[string]interface{})
		for j, h := range headers {
			hop[h] = fields[j]
		}
		hops = append(hops, hop)
	}

	return map[string]interface{}{
		"destination": strings.Split(lines[len(lines)-1], " ")[0],
		"timestamp":  time.Now().UTC().Format(time.RFC3339),
		"hops":      hops,
	}
}

func main() {
	interval := 300
	fmt.Sscanf(getEnv("MTR_INTERVAL", "300"), "%d", &interval)

	count := 10
	fmt.Sscanf(getEnv("MTR_COUNT", "10"), "%d", &count)

	outputPath := getEnv("MTR_OUTPUT_PATH", "/data/mtr")
	mtrBin := getEnv("MTR_BIN", "/usr/bin/mtr")
	rawDests := getEnv("MTR_DESTINATIONS", "1.1.1.1")

	var destinations []string
	for _, d := range strings.Split(rawDests, ",") {
		d = strings.TrimSpace(d)
		if d != "" {
			destinations = append(destinations, d)
		}
	}

	if err := os.MkdirAll(outputPath, 0755); err != nil {
		fmt.Printf("[mtr-runner] ERROR: cannot create output path: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("[mtr-runner] Starting.")
	fmt.Printf("  Interval : %ds\n", interval)
	fmt.Printf("  Count    : %d packets\n", count)
	fmt.Printf("  Output   : %s\n", outputPath)
	fmt.Printf("  Targets  : %v\n", destinations)

	for {
		for _, dest := range destinations {
			runMTR(mtrBin, dest, count, outputPath)
		}
		fmt.Printf("[mtr-runner] Cycle complete. Sleeping %ds...\n", interval)
		time.Sleep(time.Duration(interval) * time.Second)
	}
}
